app/shortcuts/shortcuts.tsx

```diff
@@ -0,0 +1,154 @@
+    this.code = code;
+  }
+
+  matches(event: KeyboardEvent): boolean {
+    return event.shiftKey == this.shiftKey && event.code === this.code;
+  }
+
+  isEqual(other: KeyCombo): boolean {
+    return this.shiftKey == other.shiftKey && this.code === other.code;
+  }
+
+  toString(): string {
+    if (this.shiftKey) {
+      return "shift-" + this.code;
+    } else {
+      return this.code;
+    }
+  }
+}
+
+class Shortcut {
+  keyCombo: KeyCombo[];
+  keyComboPosition = 0;
+  action: () => void;
+
+  constructor(keyCombo: KeyCombo[], action: () => void) {
+    this.keyCombo = keyCombo;
+    this.action = action;
+  }
+
+  matchKeyboardEvent(event: KeyboardEvent) {
+    if (this.keyCombo[this.keyComboPosition].matches(event)) {
+      this.keyComboPosition++;
+    } else {
+      this.keyComboPosition = 0;
+    }
+    if (this.keyComboPosition == this.keyCombo.length) {
+      this.action();
+      this.keyComboPosition = 0;
+    }
+  }
+
+  // Returns true if this keyboard shortcut "collides" with the provided
+  // keyboard shortcut. Collides means that one of the shortcut keystroke
+  // combinations is a prefix of the other, so it will always be activated
+  // when the longer one is entered.
+  // TODO(iain): this check is not comprehensive, we should check for
+  // collisions within another prefix, e.g. 'a' should collide with 'g-a'.
+  collidesWith(other: Shortcut): boolean {
+    let len = Math.min(this.keyCombo.length, other.keyCombo.length);
+    for (let i = 0; i < len; i++) {
+      if (!this.keyCombo[i].isEqual(other.keyCombo[i])) {
+        return false;
+      }
+    }
+    return true;
+  }
+}
+
+export class Shortcuts {
+  shortcuts: Map<string, Shortcut> = null;
+
+  // Registers the provided keyboard combination + action as a shortcut.
+  // Returns a handle that may be used to deregister this shortcut when it is
+  // no longer invokable. This function will throw an error if the same
+  // shortcut is registered multiple times, so if the calling component will
+  // unmount and remount, shortcuts should be deregistered and reregistered.
+  public register(keyCombo: KeyCombo, action: () => void): string {
+    return this.registerSequence([keyCombo], action);
+  }
+
+  // Registers the provided keyboard combination sequence + action as a
+  // shortcut. Returns a handle that may be used to deregister this shortcut
+  // when it is no longer invokable. This function will throw an error if the
+  // same shortcut is registered multiple times, so if the calling component
+  // will unmount and remount, shortcuts should be deregistered and
+  // reregistered.
+  public registerSequence(keyCombo: KeyCombo[], action: () => void): string {
+    if (this.shortcuts == null) {
+      this.shortcuts = new Map<string, Shortcut>();
+      document.addEventListener(
+        "keydown",
+        function (event: KeyboardEvent) {
+          if (window.localStorage.getItem(keyboardShortcutsKey) !== keyboardShortcutsValue) {
```

Comment:
nit: not a huge deal but I wonder if this might cause some subtle jank while typing since it's called on every keystroke -  I know localStorage used to be somewhat slow but maybe it has been optimized at this point. wdyt about referencing `preferences.keyboardShortcutsEnabled` here instead just to be safe?

Reply:
Sounds good, done